### DISCUSSION TITLE
Difference between using Message Extension query for Bot Framework v.s. Teams AI API

### DISCUSSION THREAD

**Discussion by garrytrinder** [microsoft/teams-sdk#638] [Q&A] [answered]
Scenario:

When building a Message Extension with Teams AI Library, there is an issue with accessing the search query value entered by the user.

**To Reproduce**

- Create a message extension using Azure Bot Service and Teams AI Library, the query value in the query handler is returned in `query.parameters.queryText`.
- Create a message extension using Bot Framework Portal (dev.botframework.com) and Teams AI Library, the query value in the query handler is returned in `query.parameters.searchQuery`.

**Expected behavior**

The search query value should be accessed in the same way. 

**Additional context**

The issue appears to be caused by the way in which the Teams AI Library flattens the `context.activity.value` object to enable accessing the value by name.

_Teams AI Library - Bot Framework Portal_
![image](https://github.com/microsoft/teams-ai/assets/11563347/2e2046bf-4e76-4005-9f9e-5edb97af2f62)

_Teams AI Library - Azure Bot Service_
![image](https://github.com/microsoft/teams-ai/assets/11563347/d62a00b1-0aec-407e-9e2c-4e5edcf5b6be)

When using Bot Framework SDK it is common to access the search query value using `query.parameters[0].value`.

When creating a Message Extension using Teams Toolkit, it uses Bot Framework Portal (dev.botframework.com) to register a bot, but when deployed to Azure it uses an Azure Bot Service, so this bug requires developers to code for both scenarios.

**Comment by corinagum** [comment]
Hey @garrytrinder, thanks for filing this bug. Are your bots in JS?

**Comment by garrytrinder** [comment]
Yes this is a JS/TS based bot.

**Comment by corinagum** [comment]
Hey @garrytrinder, thanks for your patience. 

To start with long story short: I moved this to the discussions section of the repo due to the fact that this is not a bug, but it's a really good conversation to have for distinguishing between using the teams-ai repo and Bot Framework. 

1. Please note that Bot Framework Dev Portal uses Azure Bot Service, so this doesn't change the behavior of deployment. The difference is that the dev portal provides UI specific towards managing bots.
2. Compare two different samples:
    1. See the link to the ["Teams Toolkit deployment" sample using message extensions](https://github.com/OfficeDev/TeamsFx-Samples/blob/v2.3.0/NPM-search-connector-M365/messageExtensionBot.js ), where the user can run 'create a bot' using the Teams Toolkit VS Code extension.
    1. See the [Teams AI Message Extension sample](https://github.com/microsoft/teams-ai/blob/main/js/samples/02.messageExtensions.a.searchCommand/src/index.ts#L76), which uses a Teams Toolkit yml file during deployment.

The key difference between these two samples is that they are using two different APIs. The first one is using the Bot Framework API, while the second is using Teams AI API.

The first sample handles message extensions by extending the `TeamsActivityHandler`:

```typescript
const { TeamsActivityHandler, CardFactory} = require("botbuilder");

class MessageExtensionBot extends TeamsActivityHandler {

  // Message Extension Code
  // Search.
  async handleTeamsMessagingExtensionQuery(context, query) {
    const searchQuery = query.parameters[0].value;
    const response = await axios.get(
      `http://registry.npmjs.com/-/v1/search?${querystring.stringify({
        text: searchQuery,
        size: 8,
      })}`
    );

    const attachments = [];
    response.data.objects.forEach((obj) => {
      const heroCard = CardFactory.heroCard(obj.package.name);
      const preview = CardFactory.heroCard(obj.package.name);
      preview.content.tap = {
        type: "invoke",
        value: { name: obj.package.name, description: obj.package.description },
      };
      const attachment = { ...heroCard, preview };
      attachments.push(attachment);
    });
...
});
```

Note that `TeamsActivityHandler` is being imported from `'botbuilder'`, not the teams-ai Library. 

The second sample uses the Teams AI API: 

```typescript
import { Application, DefaultTurnState } from '@microsoft/teams-ai';
import { createNpmPackageCard, createNpmSearchResultCard } from './cards';

// Define storage and application
const storage = new MemoryStorage();
const app = new Application({
    storage
});

// Listen for search actions
app.messageExtensions.query('searchCmd', async (context: TurnContext, state: DefaultTurnState, query) => {
    const searchQuery = query.parameters.queryText ?? '';
    const count = query.count ?? 10;
    const response = await axios.get(
        `http://registry.npmjs.com/-/v1/search?${new URLSearchParams({
            size: count.toString(),
            text: searchQuery
        }).toString()}`
    );

    // Format search results
    const results: MessagingExtensionAttachment[] = [];
    response?.data?.objects?.forEach((obj: any) => results.push(createNpmSearchResultCard(obj.package)));

    // Return results as a list
    return {
        attachmentLayout: 'list',
        attachments: results,
        type: 'result'
    };
});
```

It can be a little confusing, since Teams AI takes a dependency on `botbuilder`, but it's important to note that these are different APIs and there is no expectation for the usage to match.

Teams AI Library is intended as a wrapper for botbuilder, where accessing the Teams features that we expose is intended to be 'simpler' than if doing it in BotBuilder. As you noted, the Teams AI Library flattens the query value, and this is intentional. From the above, the Bot Framework code needs to extend the `TeamsActivityHandler` class and implement `handleMessagingExtensionQuery`, while instead teams-ai declares the app (`const app = new Application({...});)` then provides the messageExtensions module in order to set up queries. `app.messageExtensions.query('searchCmd', ...`

From the get-go, the first sample is implementing a bot that extends `TeamsActivityHandler`, but in Teams AI, we create the app, and `app.messageExtensions.query` has a different (but documented) usage.

Even though creating a bot using the Teams AI library still takes dependencies / objects from botbuilder, the implementation and usage is not guaranteed to be exactly the same, and this is a great example of how/why.

**Reply by garrytrinder** [reply]
Thanks for the reply @corinagum 

> Please note that Bot Framework Dev Portal uses Azure Bot Service,

This was my assumption, however I am seeing a subtle difference in the contents of the `value.parameters` array in the incoming message. 

If we leave Teams AI library aside for a moment and focus on the incoming message.

- The first screenshot shows that the `value.parameters` array contains a single object with the name property set to `searchQuery`.
- The second screenshot shows that the `value.parameters` array contains a single object with the name property set to `queryText`.

The only difference between the two is:

- The first screenshot is an incoming message from a bot that has been registered via `dev.botframework.com` (using TTK).
- The second screenshot is an incoming message from a bot that has been registered in Azure and uses an `Azure Bot Service`.

Given the above examples, Teams AI library will provide access to the query value entered by the user in different ways depending on how the bot was registered.

- When registered using `dev.botframework.com`, it uses `query.parameters.searchQuery`
- When registered using `Azure Bot Service`, it uses `query.parameters.queryText`

When using TTK, the default way that a bot is registered for local development is by using `dev.botframework.com`, therefore Teams AI library will return the query value using `query.parameters.searchQuery`.

If we take your example code, specifically this line `const searchQuery = query.parameters.queryText ?? '';`, then it would always return empty as `queryText` would not be returned in the parameters array, `searchQuery` would.

To add further complication, when using TTK to provision the bot to Azure, the app will then use an Azure Bot Service. 

This will result in Teams AI library using `query.parameters.queryText` when running remotely, instead of `query.parameters.searchQuery` as it does during local development. This will result in the bot not working when running in Azure which will lead to developer frustration.

The code that Teams AI library uses to flatten the `value.parameters` array is below.

https://github.com/microsoft/teams-ai/blob/abe3311f43808430e2c6db4cf9795848fa19eb30/js/packages/teams-ai/src/MessageExtensions.ts#L325-L329

Using `botbuilder` we can get around the issue by not using the named property and instead access the first element direct, e.g. `value.query.parameters[0].value`, however this is not possible using Teams AI Library.

I hope this clears things up.

**Reply by corinagum** [reply]
![image](https://github.com/microsoft/teams-ai/assets/14900841/63b0545e-df35-4a67-a343-a641c9350b8b)

do you mean the configurable `name` in `parameters`  v.s. the `queryText`? 
[Teams schema](https://learn.microsoft.com/en-us/microsoftteams/platform/resources/schema/manifest-schema)

**Reply by corinagum** [reply]
Could you share the `composeExtensions` part of both apps' manifests?

**Reply by Stevenic** [reply, marked as answer]
@garrytrinder you're correct that I'm flattening out the query parameters, but I'm only flattening them for `query` variable, not the activity. I don't touch the activity.

When I look at your two original screen shots I can tell that these are two different bots. The recipient ID's are different and the service URL's are two different. That would imply that the reason you're seeing different names is that each bots manifest used a different name. That's part of the issue with search me's. There's no standard. You can name your query params anything you want and you can also have multiple parameters which is why I flatten them out. 

I realize that its very common to have a single search parameter but that's not always the case so I tried to avoid making any assumptions around that.

**Reply by garrytrinder** [reply]
Thank you for the replies @corinagum and @Stevenic, much appreciated.

> You can name your query params anything you want and you can also have multiple parameters which is why I flatten them out.

Apologies for causing some confusion here, it wasn't clear to me that there was a direct link between the `composeExtensions.commands.parameters.name` value and that this dictated the value of the `name` property in the response from the bot. It's clear now that is the case.
